src/backend/metal/src/native.rs

```diff
@@ -219,30 +222,130 @@ impl hal::DescriptorPool<Backend> for DescriptorPool {
                 };
                 Ok(DescriptorSet::Emulated(Arc::new(Mutex::new(inner))))
             }
-            DescriptorPool::ArgumentBuffer { ref buffer, total_size, ref mut offset } => {
+            DescriptorPool::ArgumentBuffer { ref buffer, range, ref mut empty_ranges } => {
                 let (encoder, stage_flags) = match layout {
                     &DescriptorSetLayout::ArgumentBuffer(ref encoder, stages) => (encoder, stages),
                     _ => return Err(pso::AllocationError::IncompatibleLayout),
                 };
-
-                let cur_offset = *offset;
-                *offset += encoder.encoded_length();
-                if *offset <= total_size {
-                    Ok(DescriptorSet::ArgumentBuffer {
+                let mut best_range_tuple: Option<(usize, Range<NSUInteger>)> = None;
```

Comment:
This is an unfortunately large code duplication between backends. We should find a way to share the code, also take that opportunity to add unit tests for it. I think the easiest way would be, right now, to have `src/backend/aux` *module* (not crate) that is included in both backends by doing
```rust
#[path = "../aux/range_pool.rs"]
mod range_pool;
```

Reply:
Phew the unit tests were a good idea to say the least.